k2/csrc/weights.cc

```diff
@@ -68,11 +69,67 @@ WfsaWithFbWeights::WfsaWithFbWeights(const Fsa &fsa, const float *arc_weights,
     : fsa(fsa), arc_weights(arc_weights), weight_type(t) {
   if (IsEmpty(fsa)) return;
   CheckInput(fsa, arc_weights);
+  ComputeForwardWeights();
+  ComputeBackardWeights();
+}
+
+// Mohri, M. 2002. Semiring framework and algorithms for shortest-distance
+// problems, Journal of Automata, Languages and Combinatorics 7(3): 321-350,
+// 2002.
+void WfsaWithFbWeights::ComputeForwardWeights() {
+  auto num_states = fsa.NumStates();
+  forward_state_weights = std::unique_ptr<double[]>(new double[num_states]);
+  std::fill_n(forward_state_weights.get(), num_states, kDoubleNegativeInfinity);
+
+  auto add_weights_func = AddWeights();
+  const auto &arcs = fsa.arcs;
+  forward_state_weights[0] = 0;
+  for (std::size_t i = 0; i != arcs.size(); ++i) {
+    const auto &arc = arcs[i];
+    DCHECK_GE(arc.dest_state, arc.src_state);
+    auto src_weight = forward_state_weights[arc.src_state];
+    auto &dest_weight = forward_state_weights[arc.dest_state];
+
+    double r = src_weight + arc_weights[i];
+    dest_weight = add_weights_func(dest_weight, r);
```

Comment:
No:  How many times do I have to say it, put a if statement outside the for loop!
This will require a pointer to some kind of function object, which can be quite slow (e.g. may defeat prefetch).

Reply:
If you really dont want to duplicate the code, you can template Compute*Weights() on the ForwardBackwardType enum, and put the if-statement inside the for loop:
  if (i's max) { do something } else { do something else; }
Then the if-statement will be compiled out.

Reply:
mm..I am currently writing the lambda (and try to avoid duplicate lambda)...this version is not the final version for `switch` issue, I just pushed code for relaxation issue..

Reply:
.. then, of course, you'd have to have an if-statement where you invoke Compute*Weights(), so you can use the right one.    It's a bit ugly.  I would probably prefer what I have been asking for all along, which is an if-statement outside the for loop with 2 versions.  That's easier for newbies to understand.

Reply:
The way you are currently doing it with lambdas/function-objects is a definite no.  That requires at the very least a function-object lookup each time round.  It defeats optimization and lots of other things.

Reply:
I think you shoudl accept that sometimes I'll make decisions you won't like.  If you want to argue about something, why not wait till it's something where the stakes are higher?  This is just a waste of time.